src/microservices/Microservices.IsIdentifiable/Runners/IsIdentifiableAbstractRunner.cs

```diff
@@ -219,14 +240,41 @@ public void LoadRules(string yaml)
 
         // ReSharper disable once UnusedMemberInSuper.Global
         public abstract int Run();
-
+        
         /// <summary>
         /// Returns each subsection of <paramref name="fieldValue"/> which violates validation rules (e.g. the CHI found).
         /// </summary>
         /// <param name="fieldName"></param>
         /// <param name="fieldValue"></param>
         /// <returns></returns>
         protected IEnumerable<FailurePart> Validate(string fieldName, string fieldValue)
+        {
+            // every time we lookup a value increment the counter and see if we should check cache overflow
+            if(_valuesLookedUp++ % _checkEvery == 0)
+            {
+                _valuesLookedUp = 0;
+                foreach (var c in Caches)
+                {
+                    //any column with more than the maximum number of cached values should be cleared
+                    if(c.Value.Keys.Count > MaxCacheSize)
+                        c.Value.Clear();
+                }
+            }
+
+            // make sure that we have a cache for this column name
+            var cache = Caches.GetOrAdd(fieldName,(v)=>new ConcurrentDictionary<string, FailurePart[]>());
+
+            // lookup the cached value or call ValidateImpl
+            return cache.GetOrAdd(fieldValue?? "NULL",(k)=>ValidateImpl(fieldName,fieldValue).ToArray());
```

Comment:
I've not checked how this method is called, but should we be getting values that are null here?

Also probably unlikely but could `NULL` be a value we actually do receive? Might be worthwhile making this a little more unique just in case.

Reply:
ValidateImpl is IEnumerable<> so cannot return null

Reply:
If you want I can change the string to a Guid initialized on construction... your call

Reply:
Ok sure, let's leave it as-is then.